### PR TITLE
Seednode External Tor with POW and Documentation

### PR DESCRIPTION
**Pull request by shortwavesurfer2009**
This PR attempts to update the seednode deployment documentation. This PR assumes the existence of `seednode/torrc` and `seednode/externalTorDirectBind3.patch`

Edit: torrc https://github.com/boldsuck/haveno/blob/master/seednode/torrc

And externalTorDirectBind3.patch https://gist.github.com/boldsuck/5afe89c014d6e89b6aa1843f88a01299

Edit 2: @fa2a5qj3 is the author of the patch. https://gist.github.com/fa2a5qj3/57219580e7ce746aa45ac6edf4682238

### CODE REVIEW

**Comment by shortwavesurfer2009** [comment]
Done

**Comment by boldsuck** [comment]
In --hidden Service Address= port must be deleted
```
- --hiddenServiceAddress=example.onion:port\
+ --hiddenServiceAddress=example.onion\
```

**Comment by shortwavesurfer2009** [comment]
Fixed

**Comment by boldsuck** [comment]
I'll do a PR of my torrc & install_tor.sh in the next few days.
I've added comments and
HiddenServiceNumIntroductionPoints 9 # (Default: 3).
More IntroductionPoints increases accessibility but also the probability of finding the IP of the hidden service.

**Comment by shortwavesurfer2009** [comment]
I see this uses 3003 for haveno_seednode 2. This PR would change that to 2003 to be right next to haveno_seednode. @woodser what would you prefer?

**Comment by nucash-mining** [comment]
haha shortwavesurfer. It's NaturalHempsations here seeing your commits on
github come up in my email :) You get it dawg!

**Comment by boldsuck** [comment]
Yes, I think it is best to leave the original port 3003 so that existing testnet deployments are not disturbed.
--nodePort=3003, --appName=haveno-XMR_STAGENET_Seed_3003 & torrc 3003

Let's wait and see what woodser says.

In my torrc I will further adjust the values. Maxstream is still not enough.
`[WARN] Maximum streams per circuit limit reached on rendezvous circuit 3291208567 for service z47tltuwytd5icqq4hni2ammvlugp6pcwqboeu7ngawruualxjjuu3ad. Circuit has 100 out of 100 streams. Ignoring open stream request. `
^^This is seed1 to seed2

Hmm, [Dread & WhiteHouseMarket have](https://github.com/onionltd/EndGame/blob/master/torrc) HiddenServiceMaxStreams 15 but behind onionbalance.

**Comment by shortwavesurfer2009** [comment]
Can we comment out `HiddenServiceMaxStreams` or set it to 0 to see how many connections it gets? This is one I do not well understand. The PoW queue and burst make sense to me, but what this is doing is unclear to me.

**Comment by boldsuck** [comment]
> Can we comment out `HiddenServiceMaxStreams` or set it to 0 to see how many connections it gets? This is one I do not well understand. The PoW queue and burst make sense to me, but what this is doing is unclear to me.

Proof of Work (PoW) before establishing Rendezvous Circuits = to get a (the first) connection.
MaxStreams limits the established connection.
`HiddenServiceMaxStreams 0` = unlimited :-(

I wouldn't leave that out. For clients and other seed nodes, `HiddenServiceMaxStreams 25` was already good.

It's only between my two seed nodes 1-2 times a day that so many connections are established to each other.


> Edit: If nothing else, since this is just the defaults and can be easily changed, we could get this merged and then tweak it as needed. To a seed node operator, editing one number in the file and restarting the service is not going to be a large hurdle.

Yes, just wait and see what woodser says about the ports.
Note: `systemctl reload tor`  is enough for Tor to adopt the new config parameters. You only need to restart if you change ports or create a new hidden service. When reloading, Tor keeps all established connections.

**Comment by woodser** [comment]
@fa2a5qj3 Could you modify the code to work with or without an external service, so we can incorporate that change in this PR, instead of including a patch in the PR?

File: docs/deployment-guide.md [hunk 1]
```diff
@@ -79,16 +79,63 @@ Customize and deploy haveno-pricenode.env and haveno-pricenode.service to run as
...
+7. Update repositories `sudo apt update`.
+8. Install tor and tor debian keyring `sudo apt install tor deb.torproject.org-keyring`.
+9. Replace torrc `sudo mv /etc/tor/torrc /etc/tor/torrc.default` then `sudo cp seednode/torrc /etc/tor/torrc`.
+10. stop tor `sudo systemctl stop tor`.
+
+For each seed node:
+
+1. [Build the Haveno repository](#fork-and-build-haveno).
+2. [Start a local Monero node](#start-a-local-monero-node).
+3. Apply the patch `git apply /seednode/externalTorDirectBind3.patch`.
```
Comment (woodser) [inline docs/deployment-guide.md:126]: I get an error with this command.

```
woodser@woodsers-MacBook-Pro-2 haveno % git apply /seednode/externalTorDirectBind3.patch
error: can't open patch '/seednode/externalTorDirectBind3.patch': No such file or directory
```

This worked:

`git apply ./seednode/externalTorDirectBind3.patch`

**Comment by shortwavesurfer2009** [comment]
> @fa2a5qj3 Could you modify the code to work with or without an external service, so we can incorporate that change in this PR, instead of including a patch in the PR?

I am not sure if that will be possible since using the internal tor gives a different onion address than using the external tor to manage the hidden service.

**Comment by fa2a5qj3** [comment]
> @fa2a5qj3 Could you modify the code to work with or without an external service, so we can incorporate that change in this PR, instead of including a patch in the PR?

https://gist.github.com/fa2a5qj3/0f395cc1f907802379639f2f9bbee14c

This supports external tor with torrc-defined HS; internal tor with ephemeral HS via control port.

**Comment by boldsuck** [comment]
Do I understand your patch correctly? --hiddenServiceAddress as commandline or in haveno.properties and we can use external Tor for Haveno Client too?

**Comment by shortwavesurfer2009** [comment]
So, what do i need to do? I can just change externalTorDirectBind3.patch to externalTorDirectBind4.patch, but i am guessing thats not how @woodser wants it done

**Comment by boldsuck** [comment]
> So, what do i need to do? I can just change externalTorDirectBind3.patch to externalTorDirectBind4.patch, but i am guessing thats not how @woodser wants it done

I just did a rebuild. If this works as intended, nothing needs to be patched anymore. @woodser will merge. :-)

```
git apply -R /home/user/externalTorDirectBind3.patch
git apply /home/user/externalTorDirectBind4.patch
./gradlew build -x test

```

**Comment by fa2a5qj3** [comment]
https://gist.github.com/fa2a5qj3/317913bf58c740aaf63d74e35434a1e7

Retains support for Tails/Whonix users that create HS using `--torControlPort` parameter.

**Comment by preland** [comment]
> https://gist.github.com/fa2a5qj3/317913bf58c740aaf63d74e35434a1e7
> 
> Retains support for Tails/Whonix users that create HS using `--torControlPort` parameter.
> 
> [edit - don't know how to merge it into this PR, I'll create a separate one just with the patch, sorry @shortwavesurfer2009!]

I can merge this patch directly into the PR in a few hours

**Comment by boldsuck** [comment]
Do the clients only need hs-address or also port? I started with:
--hiddenServiceAddress=some.onion --nodePort=9999

I'll try externalTorDirectBind5.patch right away. (On plain Debian) xmr_mainnet.seednodes file was broken
https://github.com/boldsuck/haveno-retodev

Yesterday the idea came up. We can now also do client authorization for Onion Service.

**Comment by woodser** [comment]
Please rebase on master, removing any changes made to source files, which have been merged into master separately: https://github.com/haveno-dex/haveno/pull/1180

**Comment by woodser** [comment]
We may also want to document how to enable proof of work for the arbitrators. Any idea what that would look like?

**Comment by shortwavesurfer2009** [comment]
> We may also want to document how to enable proof of work for the arbitrators. Any idea what that would look like?

I think that would depend on how hard the arbitrators are to attack. From what I understand, they are basically just standard peers on port 9999 and it's rather easy to change their onion addresses if need be. At least this is what I have heard.

**Comment by boldsuck** [comment]
> Please rebase on master, removing any changes made to source files, which have been merged into master separately: #1180

Do you mean from reto or haveno dex?
I have master reto + #1175 and #externalTorDirectBind5.patch. Changes made to source files was only that I add addresses in xmr_mainnet.seednodes.

**Comment by boldsuck** [comment]
> We may also want to document how to enable proof of work for the arbitrators. Any idea what that would look like?

I would have to deploy an entire Haveno network but all my monerod are somewhere in the data center.
But it is always the same (Client, Seednode, Pricenode, Arb deamon). Edit `/etc/tor/torrc`


```
HiddenServiceDir /var/lib/tor/haveno_service_foo/
HiddenServicePort <Port> 127.0.0.1:<Port>
HiddenServicePort <Port> [::1]:<Port>
....
HiddenServiceDir /var/lib/tor/haveno_service_bar/
HiddenServicePort <Port> 127.0.0.1:<Port>
HiddenServicePort <Port> [::1]:<Port>
...

```
`systemctl reload tor`
Get `hs-address: cat /var/lib/tor/haveno_service_*/hostname`
Start app with cmdline: --hiddenServiceAddress=<hostname> --nodePort=<Port> or add them in haveno.properties.

With Seednodes, Pricenode & Arb deamon on one host I would create a 2nd tor-instance `tor-instance-create` for the arb to make correlation attacks more difficult.

**Comment by shortwavesurfer2009** [comment]
This is now a dead PR.

**Pull request closed without merging**